plan/newplanbuilder.go

```diff
@@ -479,29 +477,137 @@ func (b *planBuilder) extractAggFunc(sel *ast.SelectStmt) (
+		secondIsProjection := e.orderBy && e.inExpr
+		if secondIsProjection {
+			second, first = e.proj.GetSchema(), fromSchema
+		} else {
+			first, second = e.proj.GetSchema(), fromSchema
+		}
+		projCol, err := first.FindColumn(v)
+		if err != nil {
+			e.err = errors.Trace(err)
+			return inNode, true
+		}
+		if projCol == nil {
+			projCol, err = second.FindColumn(v)
+			if err != nil {
+				e.err = errors.Trace(err)
+				return inNode, true
+			}
+			if projCol == nil {
+				e.err = errors.Errorf("Can't find Column %s", v.Name)
+				return inNode, true
+			}
+			if !secondIsProjection {
+				e.addProjectionExpr(projCol)
+			}
+		} else if secondIsProjection {
+			v.Table = projCol.TblName
+			e.addProjectionExpr(projCol)
+		}
+	case *ast.SubqueryExpr, *ast.CompareSubqueryExpr, *ast.ExistsSubqueryExpr:
+		return inNode, true
+	default:
+		e.inExpr = true
+	}
+	return inNode, false
+}
+
+func (e *astColsReplacer) Leave(inNode ast.Node) (ast.Node, bool) {
+	return inNode, true
+}
+
+type gbyColsReplacer struct {
+	fields []*ast.SelectField
+	schema expression.Schema
+	err    error
+}
+
+func (g *gbyColsReplacer) Enter(inNode ast.Node) (ast.Node, bool) {
+	switch inNode.(type) {
+	case *ast.SubqueryExpr, *ast.CompareSubqueryExpr, *ast.ExistsSubqueryExpr:
+		return inNode, true
+	}
+	return inNode, false
+}
+
+func (g *gbyColsReplacer) Leave(inNode ast.Node) (ast.Node, bool) {
+	switch v := inNode.(type) {
+	case *ast.ColumnNameExpr:
+		if col, err := g.schema.FindColumn(v.Name); err != nil {
+			g.err = errors.Trace(err)
+		} else if col == nil {
+			for _, field := range g.fields {
+				if field.WildCard == nil && v.Name.Table.L == "" && field.AsName.L == v.Name.Name.L {
+					return field.Expr, true
+				}
+			}
+			return inNode, false
+		}
+	}
+	return inNode, true
+}
+
+func (b *planBuilder) rewriteGbyExprs(p Plan, gby *ast.GroupByClause, fields []*ast.SelectField) (Plan, bool, []expression.Expression) {
+	exprs := make([]expression.Expression, 0, len(gby.Items))
+	correlated := false
+	for _, item := range gby.Items {
+		g := &gbyColsReplacer{fields: fields, schema: p.GetSchema()}
+		if g.err != nil {
+			b.err = errors.Trace(g.err)
+			return nil, false, nil
+		}
+		retExpr, _ := item.Expr.Accept(g)
+		expr, np, cor, err := b.rewrite(retExpr.(ast.ExprNode), p, nil)
+		if err != nil {
+			b.err = errors.Trace(err)
+			return nil, false, nil
+		}
+		exprs = append(exprs, expr)
+		correlated = correlated || cor
+		p = np
+	}
+	return p, correlated, exprs
+}
+
 func (b *planBuilder) buildNewSelect(sel *ast.SelectStmt) Plan {
-	oldLen := len(sel.Fields.Fields)
 	hasAgg := b.detectSelectAgg(sel)
 	var aggFuncs []*ast.AggregateFuncExpr
 	var havingMap, orderMap, totalMap map[*ast.AggregateFuncExpr]int
-	if hasAgg {
-		aggFuncs, havingMap, orderMap, totalMap = b.extractAggFunc(sel)
-	}
-	// Build subquery
-	// Convert subquery to expr with plan
-	// TODO: add subquery support.
-	//b.buildSubquery(sel)
 	var p Plan
 	if sel.From != nil {
```

Comment:
Build dual plan if From is nil, otherwise build ResultSetNote, then the duplicated code can be removed.